meta/collection.go

```diff
@@ -0,0 +1,410 @@
+
+	if polarity != nil && polarity.Primary && polarity.Reversed {
+		azimuth += 180.0
+	}
+
+	// check that the value is positive
+	for azimuth < 0.0 {
+		azimuth += 360.0
+	}
+
+	return math.Mod(azimuth, 360.0)
+}
+
+// Collections decodes the stored sensor and datalogger installation
+// times and builds a slice of overlapping time spans for the given site.
+func (s *Set) Collections(site Site) []Collection {
+	var collections []Collection
+
+	for _, recorder := range s.InstalledRecorders() {
+		if recorder.Station != site.Station {
+			continue
+		}
+		if recorder.Location != site.Location {
+			continue
+		}
+
+		for _, stream := range s.Streams() {
+			if stream.Station != site.Station {
+				continue
+			}
+			if stream.Location != site.Location {
+				continue
+			}
+
+			span, ok := recorder.Span.Extent(stream.Span)
+			if !ok {
+				continue
+			}
+
+			for _, component := range s.Components() {
+				if recorder.InstalledSensor.Make != component.Make {
+					continue
+				}
+				if recorder.InstalledSensor.Model != component.Model {
+					continue
+				}
+
+				for _, channel := range s.Channels() {
+					if recorder.Make != channel.Make {
+						continue
+					}
+
+					if recorder.DataloggerModel != channel.Model {
+						continue
+					}
+					if stream.SamplingRate != channel.SamplingRate {
+						continue
+					}
+
+					var gains []Gain
+					for _, g := range s.Gains() {
+						if g.Station != stream.Station {
+							continue
+						}
+						if g.Location != stream.Location {
+							continue
+						}
+						if g.Subsource != component.Subsource {
+							continue
+						}
+						if !span.Overlaps(g.Span) {
+							continue
+						}
+						gains = append(gains, g)
+					}
+					sort.Slice(gains, func(i, j int) bool {
+						return gains[i].Span.Start.Before(gains[j].Span.Start)
+					})
+
+					var sensors []Calibration
+					for _, c := range s.Calibrations() {
+						if c.Make != recorder.InstalledSensor.Make {
+							continue
+						}
+						if c.Model != recorder.InstalledSensor.Model {
+							continue
+						}
+						if c.Serial != recorder.InstalledSensor.Serial {
+							continue
+						}
+						if !span.Overlaps(c.Span) {
+							continue
+						}
+						sensors = append(sensors, c)
+					}
+					sort.Slice(sensors, func(i, j int) bool {
+						return sensors[i].Span.Start.Before(sensors[j].Span.Start)
+					})
+
+					var dataloggers []Calibration
+					for _, c := range s.Calibrations() {
+						if c.Make != recorder.InstalledSensor.Make {
+							continue
+						}
+						if c.Model != recorder.DataloggerModel {
+							continue
+						}
+						if c.Serial != recorder.InstalledSensor.Serial {
+							continue
+						}
+						if !span.Overlaps(c.Span) {
+							continue
+						}
+						dataloggers = append(dataloggers, c)
+					}
+					sort.Slice(dataloggers, func(i, j int) bool {
+						return dataloggers[i].Span.Start.Before(dataloggers[j].Span.Start)
+					})
+
```

Comment:
don't know if I fully understand the logic, but seems the `gains`, `sensors` and `dataloggers` are not dependent on `channels`, and can be calculated outside the channels loop L155